src/NuGet.Server.Core/Infrastructure/ServerPackageStore.cs

```diff
@@ -77,45 +78,57 @@ public void Remove(string id, SemanticVersion version, bool enableDelisting)
             }
         }
 
-        public HashSet<ServerPackage> GetAll(bool enableDelisting)
+        public async Task<HashSet<ServerPackage>> GetAllAsync(bool enableDelisting, CancellationToken token)
         {
             var allPackages = new ConcurrentBag<ServerPackage>();
-            
-            Parallel.ForEach(_repository.GetPackages(), package =>
-            {
-                ServerPackage serverPackage;
 
-                // Try to create the server package and ignore a bad package if it fails.
-                if (TryCreateServerPackage(package, enableDelisting, out serverPackage))
-                {
-                    allPackages.Add(serverPackage);
-                }
-            });
+            var tasks = _repository
+                .GetPackages()
+                .Select(package => TryAddServerPackageAsync(allPackages, package, enableDelisting, token))
+                .ToList();
+
+            await Task.WhenAll(tasks);
 
             // Only return unique packages.
             return new HashSet<ServerPackage>(allPackages, IdAndVersionEqualityComparer.Instance);
         }
 
-        private bool TryCreateServerPackage(IPackage package, bool enableDelisting, out ServerPackage serverPackage)
+        private async Task TryAddServerPackageAsync(
+            ConcurrentBag<ServerPackage> allPackages,
+            IPackage package,
+            bool enableDelisting,
+            CancellationToken token)
         {
-            try
+            // Immediatelly defer work to the background thread.
+            await Task.Yield();
+
+            // Try to create the server package and ignore a bad package if it fails.
+            var serverPackage = await CreateServerPackageAsync(package, enableDelisting, token);
+            if (serverPackage != null)
             {
-                serverPackage = CreateServerPackage(package, enableDelisting);
+                allPackages.Add(serverPackage);
+            }
+        }
 
-                return true;
+        private Task<ServerPackage> CreateServerPackageAsync(
```

Comment:
I actually used this method as a forcing factor to make all callers async. I picked this location since it is most likely to invoke async methods in the future 😉. However, with the `Task.Yield()` in `TryAddServerPackageAsync` leaves the public API the same no matter what is done in this private method. So, sure.